File: HW3/angles.py
"""
@Author Kapil Sharma ks4643
@Author Dharma Teja ds3519

The purpose of this code is to find the number of right-angled triangles from a given set of coordinates.

Time Complexity: O(N^2logN)
"""


# Method to find the greatest common divisor
def find_gcd(i, j):
    while j != 0:
        i, j = j, i % j  # Using Euclidean algorithm to find GCD
    return i


# Method to perform binary search to find occurrences of a target slope
def binary_search_slope(slope_list, target_slope):
    low, high = 0, len(slope_list) - 1
    total_count = 0

    while low <= high:
        mid = (low + high) // 2
        mid_value = slope_list[mid]
        if mid_value == target_slope:
            total_count += 1
            # Handle possible duplicates to the left of the mid index
            left = mid - 1
            while left >= low and slope_list[left] == target_slope:
                total_count += 1
                left -= 1
            # Handle possible duplicates to the right of the mid index
            right = mid + 1
            while right <= high and slope_list[right] == target_slope:
                total_count += 1
                right += 1
            return total_count
        elif mid_value < target_slope:
            low = mid + 1
        else:
            high = mid - 1
    return 0


# Method to calculate slopes between all pairs of coordinates
def calculate_slopes(coordinates):
    list_of_slopes = [[] for _ in range(len(coordinates))]
    for i, point1 in enumerate(coordinates):
        for point2 in coordinates:
            if point1 != point2:
                # Calculate the differences in coordinates
                diff_y = point1[1] - point2[1]
                diff_x = point2[0] - point1[0]
                # Calculate the greatest common divisor (GCD) to simplify the slope
                gcd = find_gcd(abs(diff_x), abs(diff_y))
                # If the GCD is zero, set slope to (0, 0), else simplify the slope
                slope_tuple = (0, 0) if gcd == 0 else (diff_x // gcd, diff_y // gcd)
                list_of_slopes[i].append(slope_tuple)
        # Sort the list of slopes for each point to enable binary search
        list_of_slopes[i].sort()
    return list_of_slopes


# Main method
def main():
    number_of_points = int(input())
    coordinates = []
    # Input the coordinates
    for _ in range(number_of_points):
        x, y = map(int, input().split())
        coordinates.append((x, y))
    # Calculate slopes between all pairs of coordinates
    list_of_slopes = calculate_slopes(coordinates)
    total_count = 0
    # Iterate through each set of slopes for each point
    for slopes in list_of_slopes:
        # For each slope, find the perpendicular slopes and count occurrences
        for (diff_y, diff_x) in slopes:
            total_count += binary_search_slope(slopes, (diff_x, -diff_y))
            total_count += binary_search_slope(slopes, (-diff_x, diff_y))
    # Divide the total count by 2 since each triangle is counted twice
    print(total_count // 2)


if __name__ == "__main__":
    main()

File: HW3/Angles.java
package HW3; /**
 * @Author Kapil Sharma ks4643
 * @Author Dharma Teja ds3519
 * This program finds if there exists more than 2set of stable matching pairs
 * Time Complexity O(N^2)
 * Space Complexity O(N^2)
 *
 * We are using Gale Shapely algorithm twice to give two pairs that are stable matches
 * We compare those matches and check if they have any common pairings
 * */


import java.util.*;

public class Angles {

    public static int binarySearchCount(ArrayList<int[]> arr, int[] target) {
        int low = 0;
        int high = arr.size() - 1;
        int count = 0;

        while (low <= high) {
            int mid = (low + high) / 2;
            int[] midVal = arr.get(mid);

            if (midVal[0] == target[0] && midVal[1] == target[1]) {
                count++;
                int l = mid - 1;
                while (l >= low && arr.get(l)[0] == target[0] && arr.get(l)[1] == target[1]) {
                    count++;
                    l--;
                }
                int r = mid + 1;
                while (r <= high && arr.get(r)[0] == target[0] && arr.get(r)[1] == target[1]) {
                    count++;
                    r++;
                }
                return count;
            } else if (midVal[0] < target[0] && midVal[1] < target[1]) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return 0;
    }

    public static int gcd(int a, int b) {
        while (b != 0) {
            int temp = a;
            a = b;
            b = temp % b;
        }
        return a;
    }

    public static List<List<Integer>> calculateSlopes(List<int[]> points) {
        List<List<Integer>> slopes = new ArrayList<>(points.size());
        for (int i = 0; i < points.size(); i++) {
            slopes.add(new ArrayList<>());
            for (int j = 0; j < points.size(); j++) {
                if (i != j) {
                    int dy = points.get(j)[1] - points.get(i)[1];
                    int dx = points.get(j)[0] - points.get(i)[0];
                    int g = gcd(Math.abs(dx), Math.abs(dy));
                    if (g == 0) {
                        slopes.get(i).add(0);
                        slopes.get(i).add(0);
                    } else {
                        slopes.get(i).add(dx / g);
                        slopes.get(i).add(dy / g);
                    }
                }
            }
            Collections.sort(slopes.get(i));
        }
        return slopes;
    }

    public static int check(List<int[]> a) {
        ArrayList<int[]>[] slopes = new ArrayList[a.size()];
        for (int i = 0; i < a.size(); i++) {
            slopes[i] = new ArrayList<>();
            for (int[] p2 : a) {
                if (!Arrays.equals(a.get(i), p2)) {
                    int dy = p2[1] - a.get(i)[1];
                    int dx = p2[0] - a.get(i)[0];
                    int g = gcd(Math.abs(dx), Math.abs(dy));
                    int sdx = g == 0 ? 0 : dx / g;
                    int sdy = g == 0 ? 0 : dy / g;
                    slopes[i].add(new int[]{sdx, sdy});
                }
            }
            slopes[i].sort((o1, o2) -> {
                if (o1[0] != o2[0]) return Integer.compare(o1[0], o2[0]);
                return Integer.compare(o1[1], o2[1]);
            });
        }

        int count = 0;
        for (ArrayList<int[]> s : slopes) {
            for (int[] slope : s) {
                count += binarySearchCount(s, new int[]{ slope[0], -slope[1] });
                count += binarySearchCount(s, new int[]{ -slope[0], slope[1] });
            }
        }

        return count / 2;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        List<int[]> points = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            int x = scanner.nextInt();
            int y = scanner.nextInt();
            points.add(new int[]{x, y});
        }
        System.out.println(check(points));
        scanner.close();
    }
}
